### DISCUSSION TITLE
Exercise 2, Question 5

### DISCUSSION THREAD

**Discussion by ewhittemore987** [LiNk-NY/IntroR#38] [General]
@ralythompson24 and classmates:
I'm sure I'm missing something quite simple, but I've been trying for a while now and wondering if anyone can help. I even tried several different variations, but all keep coming up with NULL instead of the answer. Anyone have any suggestions?
<img width="576" alt="Exercise 2, Question 5" src="https://github.com/CUNYSPHCode/IntroR/assets/171465841/ddb19303-ef7b-4cd2-ba2a-16e7b489a9e3">
<img width="576" alt="Exercise 2, Question 5" src="https://github.com/CUNYSPHCode/IntroR/assets/171465841/ddb19303-ef7b-4cd2-ba2a-16e7b489a9e3">
<img width="576" alt="Exercise 2, Question 5" src="https://github.com/CUNYSPHCode/IntroR/assets/171465841/ddb19303-ef7b-4cd2-ba2a-16e7b489a9e3">

**Comment by camilayepes16** [comment]
Hi! This is so strange. I definitely inputted nrow(mtcars) and it gave me 32. Are you running R or Python? Apparently google colab is in python by default and you have to switch it to R using the tab "runtime" > "Change runtime type" - maybe that's why it's wonky for you. Otherwise I am unsure.

**Reply by ewhittemore987** [reply]
Yes, I am running R. Everything else runs fine. The only one I can't get to run properly is the nrow(), ncol(), rownames() commands for some reason. Maybe I'll try exiting out and going back in as it seems some have had success with doing that. Not sure what else to do as it appears to be written correctly (I think...)

**Reply by LiNk-NY** [reply]
Make sure that you have the right object type for the functions that you are using. You can verify with: 
```r
class(mtcars)
```
which should return `data.frame` if it doesn't, circle back to the code where you thought you were loading `mtcars`, i.e., `data("mtcars", package = "datasets")`

**Comment by KrisDelaPena** [comment]
Hello! Can you try to run the code again but this time instead of using "mtcars <- data(mtcars)", just use "data(mtcars)" to load the data.

**Reply by ralythompson24** [reply]
@ewhittemore987  this would be my recommended solution. I think something weird happens when you save `data(mtcars)` to the object `mtcars` since technically running `data(mtcars)` is already creating the `data.frame` for you.

**Reply by LiNk-NY** [reply]
FYI: `?data` shows that the return value is: 

```
     A character vector of all data sets specified (whether found or
     not)
```

**Reply by ralythompson24** [reply]
Right! So when you run `data("mtcars")` it loads `mtcars` into your current R session. But, if you try to run `mtcars <- data("mtcars")` the object `mtcars` is a character vector, not a `data.frame` object.

**Reply by KrisDelaPena** [reply]
Oh. Is that why the results came out as NULL? Because when we use mtcars <- data("mtcars"), the dataset mtcars now only includes the value "mtcars" in it and there is no columns and rows?

**Reply by ralythompson24** [reply]
Yes, you're correct. The function `nrow()` in R is designed to return the number of rows in a matrix or a data frame. When you pass a vector to `nrow()`, it returns `NULL` because a vector does not have a dimension attribute that defines rows and columns. A vector is considered to have only one dimension, its length, regardless of whether it’s a single value or multiple values.

If you want to find the length of a vector, we would instead use `length()`

**Reply by KrisDelaPena** [reply]
In the case of the exercise homework, is it okay if I used length(mtcars) instead of ncol(mtcars) to answer the question on the number of columns? I tried both and both yielded the same results.

**Reply by ralythompson24** [reply]
Yep! That is also fine. You should recognize though, that since `data.frame` objects are a special type of `list`, the `length` function in this case is returning for you the number of items in the `list`. In the case of the `data.frame`, it would be the number of vectors contained within the `data.frame` (what we also call the "columns").
